presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +229,157 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
+     */
+    public List<Long> zOrderSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        List<Long> addresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(minimumValues, maximumValues);
+
+        Iterator<ZAddressRange> rangeIterator = addressRanges.iterator();
+        while (rangeIterator.hasNext()) {
+            ZAddressRange range = rangeIterator.next();
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                addresses.add(address);
+            }
+        }
+
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    public List<ZAddressRange> zOrderCondensedSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        ZOrderCurveSearcher curveSearcher = new ZOrderCurveSearcher(minimumValues, maximumValues);
+        return curveSearcher.findAddresses();
+    }
+
+    /**
+     * Provides a structure to search for addresses within given value ranges.
+     */
+    private class ZOrderCurveSearcher
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+        private final int[] dimensions;
+        private LinkedList<ZAddressRange> addressRanges;
+
+        /**
+         * Class constructor specifying the value ranges and initializing <code>dimensions</code>.
+         * <p/>
+         * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+         */
+        public ZOrderCurveSearcher(List<Integer> minimumValues, List<Integer> maximumValues)
+        {
+            this.minimumValues = minimumValues;
+            this.maximumValues = maximumValues;
+
+            // Create mutable version of encodingBits
+            List<Integer> bitLengths = new ArrayList<>(encodingBits);
+            this.dimensions = new int[maxBitLength];
+            for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+                for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                    if (bitLengths.get(bitLengthIndex) > 0) {
+                        dimensions[dimensionIndex]++;
+                        bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                    }
+                }
+            }
+        }
```

Comment:
If we do `for (int bitLengthIndex = 0; ...; ) { for (int dimensionIndex = ..; ;) {...}}`, do you think if we still need the "if" statement?

Reply:
Yes, we'd still need the "if" statement because each value in `bitLengths` is different based on the input `encodingBits`. 
For example, if `bitLengths` was `{1, 2, 3}`, we would loop a total of `maxBitLength = 3` times over bitLengths. 
With no "if" statement, the resulting `bitLengths` would be `{-2, -1, 0}` and dimensions would be `{2, 2, 2}` instead of `{1, 2, 2}`.